### PR TITLE
Delayed icon generation

### LINKED ISSUE #9729: FR: lazy generated icons
Currently with the new icon service we generate all thumbnail sizes based on elgg_get_icon_sizes when the image is uploaded.

This is an issue in some cases:
- If, after upload, a new icon size is added (for example in a new theme)
- If an icon size is never used

What if we only upload the master image and when a thumbnail is required, but is not yet available, and the size is present in elgg_get_icon_sizes, we generate it on demand? That will fix both issues mentioned above.

### PR DESCRIPTION
**Pull request by jeabakker**
Only master will be generated during upload, the rest on demand

Also adjusted master size to 2048x2048

ToDo
- [x] docs
- [x] check tests
- [x] fix avatar cropper
- [x] find related issues

Fixes #9729

### CODE REVIEW

**Comment by hypeJunction** [comment]
I am not sure what this is supposed to delay and why

File: engine/classes/Elgg/EntityIconService.php [hunk 1]
```diff
@@ -185,81 +185,132 @@ public function saveIconFromElggFile(ElggEntity $entity, ElggFile $file, $type =
...
+		$y1 = (int) elgg_extract('y1', $coords);
+		$x2 = (int) elgg_extract('x2', $coords);
+		$y2 = (int) elgg_extract('y2', $coords);
+		
+		$cropping_mode = ($x2 > $x1) && ($y2 > $y1);
+		
+		$sizes = $this->getSizes($entity->getType(), $entity->getSubtype(), $type);
 		foreach ($sizes as $size => $opts) {
+			
+			if (!empty($icon_size) && ($icon_size !== $size)) {
```
Comment (jdalsem) [inline engine/classes/Elgg/EntityIconService.php]: generate icon will report true if requested size is not in sizes.. is this correct?

**Review by jdalsem — APPROVED** [review]
besides the single bug, this is wonderful

**Comment by hypeJunction** [comment]
Can someone explain why this is wonderful? I fail to see the point

**Comment by jeabakker** [comment]
Advantages of delayed icon creation
- less time spent on initial upload
- unused icons don't get stored
- adding a new iconsize later will generate this when requested

File: engine/classes/Elgg/EntityIconService.php [hunk 2]
```diff
@@ -185,81 +185,132 @@ public function saveIconFromElggFile(ElggEntity $entity, ElggFile $file, $type =
 	 */
 	public function saveIcon(ElggEntity $entity, ElggFile $file, $type = 'icon', array $coords = []) {
 
+		$type = (string) $type;
+		if (!strlen($type)) {
```
Comment (hypeJunction) [inline engine/classes/Elgg/EntityIconService.php]: `strlen()` on a non-string value will generate an error. Just check if `empty`

File: engine/classes/Elgg/EntityIconService.php (hunk 1 above)
Comment (hypeJunction) [reply engine/classes/Elgg/EntityIconService.php]: You don't need to loop, just extract `$icon_size` from `$sizes` array and bail if not set. Note that `original` might be an empty array.

**Review by hypeJunction — COMMENTED** [review]
Not crazy about this, as it adds layers of untestable conditions.

File: engine/classes/Elgg/EntityIconService.php (hunk 2 above)
Comment (jeabakker) [reply engine/classes/Elgg/EntityIconService.php]: thats why i cast it to a string 1 line above

File: engine/classes/Elgg/EntityIconService.php (hunk 1 above)
Comment (jeabakker) [reply engine/classes/Elgg/EntityIconService.php]: if `$icon_size` is empty all icons will be generated. Also i added a check that if `$icon_size` is given it has to exist in the allowed sizes.

**Comment by jdalsem** [comment]
> as it adds layers of untestable conditions.

like what?

File: engine/classes/Elgg/EntityIconService.php (hunk 2 above)
Comment (hypeJunction) [reply engine/classes/Elgg/EntityIconService.php]: `(string) false` is `0`; `strlen('0')` is `1`

**Comment by hypeJunction** [comment]
> like what?

- How do you test the behavior of `hasIcon()`? A logical way to approach this would be to test if `hasIcon` before generating one? What this does is creates an icon to then test if icon exits

- What's the point of preserving master? Either delete all icons or delete none, or add a method to delete an icon of specific size. Adding half-processes doesn't help testing.

- What happens if resize fails? Current behavior is to refuse to store an icon if it can't be resized to one of the defined sizes. With this, every time you call `getIconURL()` it's going to attempt to recreate an icon, and fail over and over.

I see many points of failure, which are neither documented nor tested, so noone knows what the expected behavior should be.

**Comment by jeabakker** [comment]
> How do you test the behavior of hasIcon()? A logical way to approach this would be to test if hasIcon before generating one? What this does is creates an icon to then test if icon exits

I added a test to check if delayed creation works. other test already cover `hasIcon()` in detail

> What's the point of preserving master? Either delete all icons or delete none, or add a method to delete an icon of specific size. Adding half-processes doesn't help testing.

this is done during cropping, because cropping is based on master icon

**Pull request merged**